docs/Questions.md

```diff
@@ -0,0 +1,98 @@
+<H1> Folder structure(Pages, Components, Templates, Organisms, Molecules, Atoms) </H1>
+We should stick to one way of organizing our folder and component's structure.
+Utils have files with .tsx file extension, which I think is inappropriate.
+We have validation.js and wdyr.js files at the top level of src
+
+<H1> Styles </H1>
+Currently we both use Styled-Components and scss across our codebase. I think we should stick to one and remove the other
+
+---
+
+Mike's comment: _Personally, I prefer styled-components because they allow you to pass variables to your styles from the component_
+
+<H1> Data fetching & selectors in components </H1>
+We have a lot of ambiguity about how we interact with data in our app.
+For example:
+
+- `useConnects` in components with different ways of solving the issue, when there is no needed data(like a venue id)
+- firestore direct pushes in `actions` and components
+
+<H1> Merging flow </H1>
+We should have a well-defined merging flow to avoid misunderstandings.
+
+---
+
+Mike's comment:
+
+Depending on the CI\CD strategy we choose, we could end up with 2 flows(normal | hotfix).
+
+- Normal: feat|impr|fix branch => Staging => Production
+- Hotfix: hotfix branch => Prod && hotfix branch => Staging
+
+<H1> Branch and commit naming </H1>
+Well-defined rules for this can help keep everything tidy and understandable by all parties
+
+---
+
+Mike's comment:
+
+I suggest having only 4 allowed terms in commits, branches, PR names:
+
+- Improvement | IMPR - Refactoring
+- Fix | FIX - bug fix
+- Hotfix | HOTFIX - Direct bugfix to production(hopefully, the rarest one)
+- Feature | FEAT - New feature / Behavior change of an old feature
+
+Example:
+
+- _[GITHUB_BOARD_TICKET_NUMBER]/['feat' | 'fix' | 'impr' | 'hotfix']/[short branch description] as an example of branch naming. Should be solidified as a regexp rule_
+- _[GITHUB_BOARD_TICKET_NUMBER]: [Commit extensive description in present simple] - commit naming example. Should be solidified as a regexp rule_
+
+<H1> Merging vs rebasing </H1>
```

Comment:
_I vote for rebasing as it helps in keeping the history clean. Also, one should rebase his branch with staging every day to avoid having lots of conflicts. After we discuss this one, I could write the perfect flow of working on a feature(As I see it)_

Reply:
In the past I have personally preferred rebasing as well, but it seems to have a negative impact on GitHub PR review workflows, as a rebase means that you can't use features such as 'changes since last review', etc.

![image](https://user-images.githubusercontent.com/753891/103854548-d29aa000-5104-11eb-9100-b46052ae8460.png)

Since we squash-merge our PR's when merging into `staging`, having a 'perfectly clean commit history' isn't really as important as it would be if we weren't squash merging.

Merging `staging` -> `master` means that we have a canonical list of deploys that were made, that references back to the PR's they were made in. And then based on the standard way that GitHub adds the PR number to the end of a commit when squash merging, we also have a clean mapping back to the original PR.

I would potentially be open to us being able to use rebasing when merging `staging` -> `master`, but that would mean that we wouldn't be able to easily merge hotfixes back into `staging`, as the branches will be divergent. That's what stopped us doing it in the past.

Personally, i'm actually really not a fan of the whole 'merge `hotfix` to `master` then backport to `staging`' approach. It feels like a hack around a system that we could just design better from the beginning; and it doesn't really align well with the CI/CD model.

Removing the concept of `staging` as a branch, and moving towards continuous delivery (where we can't get 'choked up' by code in `staging` that isn't ready to be deployed) is my preferred solution. As part of this, we would also be leaning more heavily on patterns that allow us to 'not break things' for users, namely using feature flags, and patterns such as major changes being deployed in `v2` components side-by-side with the originals, etc.

----

In the future when we move to more of a continuous-delivery sort of pattern, I would actually like to remove the difference we have between the `staging` branch and the `master` branch. Individual PR's are used for developing features, when they are merged, they should be guaranteed to not break anything, therefore `master` should always be deployable.

Having the `staging` and then needing to 'deploy' it to master is sort of convoluting different concepts in my mind. Code is code. Deploying code to a production environment should be handled in a better way, not by requiring indirection in our branches like we currently have it.

In a previous workplace we raised PRs against `master`. Our CI/CD would automatically deploy that to our staging environment, and then there was a button in our CI workflow that we could press (after testing in `staging`) that would release that exact same code into `production`.

Our current workflows technically mean we are rebuilding the code before it gets deployed to production, so while it should be the same, it isn't technically the exact same code we were testing in staging.

I expect when we add the 'per-PR staging deployment' feature (see https://github.com/sparkletown/internal-sparkle-issues/issues/108) that this will become a lot more practical/closer to being able to achieve, as right now we sort of rely on merging to `staging` for non-tech team members to be able to play with/review/QA things; which really should be happening before the PR is approved/merged.

Reply:
See also: https://github.com/sparkletown/internal-sparkle-docs/pull/1#discussion_r579889314